File: app/src/main/java/com/marvelcomics/brito/marvelcomics/application/injection/module/RepositoryModule.java
package com.marvelcomics.brito.marvelcomics.application.injection.module;

import com.marvelcomics.brito.data.datasource.remote.MarvelWebService;
import com.marvelcomics.brito.data.repository.character.CharacterRepository;
import com.marvelcomics.brito.data.repository.character.RemoteCharacterRepository;
import com.marvelcomics.brito.data.repository.comics.ComicRepository;
import com.marvelcomics.brito.data.repository.comics.RemoteComicRepository;
import com.marvelcomics.brito.data.repository.series.RemoteSeriesRepository;
import com.marvelcomics.brito.data.repository.series.SeriesRepository;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class RepositoryModule {

    @Provides
    @Singleton
    static CharacterRepository providesRemoteCharacterRepository(MarvelWebService marvelWebService) {
        return new RemoteCharacterRepository(marvelWebService);
    }

    @Provides
    @Singleton
    static ComicRepository providesRemoteComicRepository(MarvelWebService marvelWebService) {
        return new RemoteComicRepository(marvelWebService);
    }

    @Provides
    @Singleton
    static SeriesRepository providesRemoteSeriesRepository(MarvelWebService marvelWebService) {
        return new RemoteSeriesRepository(marvelWebService);
    }
}

File: app/src/main/java/com/marvelcomics/brito/di/RepositoryModule.kt
package com.marvelcomics.brito.di

import com.marvelcomics.brito.data.repository.characters.CharacterRepository
import com.marvelcomics.brito.data.ktor.ComicKRepository
import com.marvelcomics.brito.data.repository.comics.ComicRepository
import com.marvelcomics.brito.data.repository.series.SeriesRepository
import com.marvelcomics.brito.domain.repository.ICharacterRepository
import com.marvelcomics.brito.domain.ktor.IComicKRepository
import com.marvelcomics.brito.domain.repository.IComicRepository
import com.marvelcomics.brito.domain.repository.ISeriesRepository
import org.koin.dsl.module

object RepositoryModule {
    val repositories = module {
        factory<ICharacterRepository> { CharacterRepository(get(), get()) }
        factory<IComicRepository> { ComicRepository(get(), get()) }
        factory<IComicKRepository> { ComicKRepository(get(), get(), get()) }
        factory<ISeriesRepository> { SeriesRepository(get(), get()) }
    }
}
